### ISSUE TITLE
Minecraft 1.17: Update tracking issue

### ISSUE THREAD

**Issue by Proximyst** [PaperMC/Paper#5785]
# Step -1 - Tracking the update process

This acts as a tracker for all issues and progress for the update to Minecraft 1.17.

Preliminarily, these are the points we will need to fulfill:

- [x] Spigot 1.17 release
- [x] Potentially update build scripts
- [x] Rebase API patches (over 300), resolving conflicts
- [x] Rebase Server patches (almost 800), resolving conflicts, preliminarily remove patches we believe are unnecessary
- [x] ~~Fix compilation issues from previous steps~~
- [x] Finish remaining patches which were initially skipped
- [x] Fix server crash issues and other major issues discovered before test builds are released
- [x] Release regular test builds and resolve other critical issues
- [x] Once significant issues are fixed and no new significant issues come up for some amount of time (undecided), mark builds as stable and remove warnings about them being test builds

Any Paper team member may change the status of these, and more in-depth or important updates besides the milestones may be entered in the comments.

#### Spam comments such as asking for an ETA _will_ be removed and you _may_ be blocked from the organisation.

---

Like with any other Minecraft release, we get increasing volumes of questions on when Paper will be available. With that, we have a lot of work ahead of us -- all of which will be laid out in the following paragraphs. If you’d like to skip the reading and hop right into watching us do it live in the PaperMC voice channels, feel free to do so after the Spigot release.

# Step 0 - Blocked on Spigot
Because Paper is a fork of Spigot, we have to wait until Spigot releases their 1.17 builds before any work on our end can start. Spigot has waited until the full major release for the previous two versions; therefore, the 1.17 Spigot release is our first milestone.

# Step 1 - Preliminary work 
Upon Spigot’s 1.17 release, we first need to ensure the build system will remain functional. Paper’s build system is prone to breakage at pretty much any major change from either Mojang or Spigot, and with this in mind, we will need to be thorough and analyze their changes (among which is likely the Java 16 bump).

# Step 2 - Updating the API
Once we complete fixing the build system, our first task is to work on our 300+ API patches. In the best-case scenario, each patch applies automatically, however in the worst-case, we need to do a tonne of manual work. Most of the time, our various tools can deal with most of the changes and we are left to do the clean-up.

# Step 3 - Updating the server
As soon as the API has been updated, we will work on the server itself by updating the nearly 800 patches we make to the server implementation. This follows the same process of updating the API, however with major changes of both Mojang and Spigot code during major updates in mind, our patches often require manual work in resolving issues. Additionally, during this process, we may find that some patches need to be dropped entirely because they are either unnecessary (such as bugs fixed upstream) or the code has changed enough to make the only viable option a careful rewrite later on. This is by far the most time-consuming step, and will likely take multiple days of ‘round-the-clock work by the PaperMC team and its lovely contributors.

# Step 4 - Get it to compile
Once we finish updating the server, we then start to try to compile the server. This will always fail on the first try as some minor mistakes will occur in the previous steps.

# Step 5 - Squash all the bugs
The final step is testing; most of the time the server will crash instantly due to behavioural changes in the server. This begins a cycle of testing, finding bugs, and fixing. Once all the immediately obvious bugs are fixed, we will likely open up a small server to have a few people help us test further to find more bugs. Once we get to the point where we feel we’ve tested our current progress enough, we will provide public test builds. These builds will still have bugs and we recommend you only run these on disposable test servers. If you do run these builds, be sure to report any issues you encounter, especially with plugins and game mechanics. They will often be fixed rather quickly. Once the rate of new issues slows down, it will be merged into the main branch and made our primary release such that everybody can download Paper 1.17 from our website. However, be warned, these are still very early builds and there will be dragons.

If any part of the Paper updating process sounds interesting to you, be sure to check in with us on our live-streams, in which several team members and contributors will stream themselves working on this project in turn. Join us, hang out with us, ask questions, and provide a little company for us on this often pretty monotonic and boring work.

---

#### Spam comments such as asking for an ETA _will_ be removed and you _may_ be blocked from the organisation.

**Comment by Proximyst** [comment]
**Milestone**: Minecraft 1.17 has been released. We are still awaiting the release of Spigot for Minecraft 1.17.

**Comment by Proximyst** [comment]
**Milestone**: Spigot 1.17 has been released. We are now commencing work on Paper for 1.17.

**Comment by MiniDigger** [comment]
~~todo: @MiniDigger | Martin log4j2 version needs bumping in pom.xml still, bom~~ ✔️

**Comment by MiniDigger** [comment]
headsup: due to build changes by spigot, the team has decided to revamp our build tools from the ground up. this does sound scary, but we have been working on this for a year already and are prepared, just need some finishing touches.

**Comment by kashike** [comment]
~~todo: Change maven packages to `io.papermc` base~~ ✔️

**Comment by Proximyst** [comment]
To do: update BungeeCord chat API to 1.17? Do we care?

**Comment by Proximyst** [comment]
To do: Move KeyedObject, ServerWorkerThread, MCUtil

**Comment by MiniDigger** [comment]
~todo: move patches into patches/api and patches/server~ :heavy_check_mark:

**Comment by kashike** [comment]
~todo for @kashike: re-add packet fields in adventure~ #5844

**Comment by Proximyst** [comment]
To do: Remove Aikar's cleaner compat util; we have the JDK 9+ classes without compromise now.

**Comment by Proximyst** [comment]
To test: Ensure expired temp bans work correctly. cc @mrfloris

**Comment by MiniDigger** [comment]
check that "Configurable mob spawner tick rate" was applied correctly

**Comment by MiniDigger** [comment]
todo: figure out why "comment out useless overrides" in mcutil patch, blockpos class is suddenly needed now

**Comment by jpenilla** [comment]
todo: sending resource packs now has a prompt Component, api should be updated eventually cc @Machine-Maker

**Comment by MiniDigger** [comment]
todo readd https://github.com/PaperMC/Paper/pull/5797/files#diff-7cd43992eca19019d4121a4c1a8689addd5f1df32ebd205bef833c8aa62bbef7L125

**Comment by JRoy** [comment]
todo: check custom player collisions patch to make sure it works (do we even want that anymore)

**Comment by MiniDigger** [comment]
~todo: check if title timings still accept -1 (string based action bar api)~ mini is dum dum

**Comment by MiniDigger** [comment]
~~todo: explanation for build-data/mcdev-imports.txt:~~
```
# You can use this file to import files from vanilla into the project
# both fully qualified and a file based syntax are accepted here:
# net.minecraft.world.level.entity.LevelEntityGetterAdapter
# net/minecraft/world/level/entity/LevelEntityGetter.java
```

**Comment by MiniDigger** [comment]
~~todo: check if we still want "Zs dum patch" (Add system property to disable book size limits)~~ moved out

**Comment by aurorasmiles** [comment]
~~todo: check if we still need "fix dragon server crashes" - might be fixed in vanilla~~
no we don't

**Comment by electronicboy** [comment]
~~https://github.com/FabricMC/tiny-remapper/tree/remapbridges - "fixes" (Needs testing) what appears to be borking log4j in a spigot mapped jar~~

**Comment by aurorasmiles** [comment]
~~todo: check if we still need "fix exploit that allowed creating colored signs"~~

**Comment by electronicboy** [comment]
~~work on kernel modules required for wingdings console support~~

**Comment by DenWav** [comment]
~~todo: move project and package version info to Paper-Server module rather than in parent~~

**Comment by electronicboy** [comment]
~~port: https://github.com/PaperMC/Paper/pull/5793~~
PR'd

**Comment by DenWav** [comment]
todo: check Improve death events patch

**Comment by DenWav** [comment]
~~todo: check `PiglinAi` to `PiglinAI` conversion on Windows from forks issue~~

**Comment by DenWav** [comment]
~~(minor) todo: check server size increase (20MB increase from 1.16.5)~~
This appears to have been fixed

**Comment by DenWav** [comment]
todo: implement userdev paperweight plugin

**Comment by DenWav** [comment]
~~todo: investigate index file issue with gradle daemon and `clone<upstream>` task in paperweight forks~~

If we figure out steps to reliably repro this then we can look into it. Until then, running `./gradlew --stop` whenever this comes up fixes it.

**Comment by DenWav** [comment]
~~todo: look into patch remap for forks~~

Added `--skip-remapping-patches` argument to remap patches task so forks can use it to skip remapping paper patches and just remap their own.

**Comment by kennytv** [comment]
~~todo: see where `0100-Avoid-blocking-on-Network-Manager-creation` went (and then apply diff in `ServerConnectionListener` in `Optimize-Network-Manager-and-add-advanced-packet-sup.patch`, see the old remapped patch)
edit: someone dropped the patch on accident, so it has to be readded + the other step mentioned~~

**Comment by kennytv** [comment]
~~todo: when `0243-Add-Debug-Entities-option-to-debug-dupe-uuid-issues` is re-added, move the added `tracker` field out of this patch: <https://github.com/PaperMC/Paper/commit/d42140148f655a3a3d70c69979ecc03ca34b8af1#diff-47a662ae0531e46ecbf99f48d6042284fe1e79fc8927baa883012f312979afde>~~

**Comment by aurorasmiles** [comment]
~~todo: look at~~

**Comment by MiniDigger** [comment]
todo: check if fix-stuck-in-sneak-when-changing-worlds is still needed (mojira issue is closed)

**Comment by zachbr** [comment]
> 
> 
> todo: check if we still want "Zs dum patch" (Add system property to disable book size limits)

You probably don't. It was from a different time before books were as jank as they are now. As time has gone on they've only proven to be worse and worse.

**Comment by electronicboy** [comment]
~~Bees-get-gravity-in-void.-Fixes-MC-167279.patch - needs min height change~~

**Comment by JRoy** [comment]
`Don't mark null chunk sections for block updates` needs to be checked, entire system around it has changed. May not even be needed

**Comment by MiniDigger** [comment]
allow delegration to vanilla chunk gen is most certainly broke (like always)

**Comment by MiniDigger** [comment]
~~deploy paperclip to paper repo (via action, @DenWav), then change repo in build.gradle.kts~~

**Comment by aurorasmiles** [comment]
Milestone: We rebased all patches that we did not temporarily remove. The next step is now to re-add those removed patches. Those patches were initially skipped because they require a bit more in-depth examination. Some patches were removed because they are no longer necessary, those will not be added back.

**Comment by kennytv** [comment]
todo: looked fine on brief testing of the 2 modes and disabled, but check if something has been missed in https://github.com/PaperMC/Paper/blob/eab6d9f53323bad608c7597ce9152a18f2e6ff15/patches/server/0364-Anti-Xray.patch just to be sure ~~(probably didn't account for negative min y height?)~~ it works with custom world height now, but still need to make it able to obfuscate blocks below y=0

**Comment by kennytv** [comment]
note on `0438-Workaround-for-Client-Lag-Spikes-MC-162253`: The ceilingLightMask method is too cursed to figure out what *exactly* is supposed to be met there, but this is updated to likely match the previous behavior: https://paste.gg/p/KennyTV/fb71550e4c0e4909b487b74efeb3d09f
I didn't obverse any difference in lag or no lag (and am not 100% sure it's correct), so I didn't pull it.

**Comment by electronicboy** [comment]
TODO: incremental chunk saving does not seem to care about entities

**Comment by DenWav** [comment]
## FOR DEVELOPMENT TESTING ONLY

Current test build: <removed>

This is *NOT* a statement of when Paper 1.17 will be released. This jar is *ALPHA SOFTWARE*. *DO NOT USE THIS FOR YOUR SERVER*.

This is useful for those who want to help us test before we release actual test builds. This is useful for plugin developers who want to test compatibility.

*DO NOT USE THIS FOR YOUR SERVER UNDER ANY CIRCUMSTANCES.*

**Comment by kennytv** [comment]
~~Fix this ![aaa](https://i.imgur.com/7xem8LD.png) already reported with the first test build above~~

**Comment by DenWav** [comment]
## FOR DEVELOPMENT TESTING ONLY

Experimental build 3: <removed>

This is *NOT* a statement of when Paper 1.17 will be released. This jar is *ALPHA SOFTWARE*. *DO NOT USE THIS FOR YOUR SERVER*.

This is useful for those who want to help us test before we release actual test builds. This is useful for plugin developers who want to test compatibility.

*DO NOT USE THIS FOR YOUR SERVER UNDER ANY CIRCUMSTANCES.*

**Comment by electronicboy** [comment]
TODO: Investigate if aikars "catch all" dupe patch is still entirely needed

There are likely still parts of that patch which are relevant, but, in 1.17, mojang added a removal cause to entities which marks dead entities in a manner that they should not process death drops, etc, when removed for specific causes; This patch has caused one or two issues but has generally saved our asses before many various dupes have been found, however, at least the entity aspects, may actually be no longer needed

**Comment by DenWav** [comment]
1.17.1 is released and we consider it stable. All issues related to 1.17 will be handled as a standard Paper issue now, rather than targeting a 1.17 release.
